all: fix all undefined doxygen groups

aabadie opened the pull request:
<!--
The RIOT community cares a lot about code quality.
Therefore, before describing what your contribution is about, we would like
you to make sure that your modifications are compliant with the RIOT
coding conventions, see https://github.com/RIOT-OS/RIOT/wiki/Coding-conventions.
-->

### Contribution description

This PR fixes #8972 by adding a doxygen groups checker script. This script is based on the [gist](https://gist.github.com/kaspar030/78b6a26e710a03ae1d28cae5c38f3cbe) provided by @kaspar030 and which is working very well.
I slightly adapted it to not take into account the vendor header files and to make it return the full list of non matching groups.

In the meantime, I fixed all issues raised by the script. But I think I'll split this PR in several parts to simplify the reviews:
1. provide the check script alone, e.g not integrated in the CI
2. provide a PR per top folder (boards, cpu, etc)
3. enable the script in the CI

<!--
Put here the description of your contribution:
- describe which part(s) of RIOT is (are) involved
- if it's a bug fix, describe the bug that it solves and how it is solved
- you can also give more information to reviewers about how to test your changes
-->


### Issues/PRs references

fixes #8972 ~and based on #9266~

<!--
Examples: Fixes #1234. See also #5678. Depends on PR #9876.

Please use keywords (e.g., fixes, resolve) with the links to the issues you
resolved, this way they will be automatically closed when your pull request
is merged. See https://help.github.com/articles/closing-issues-using-keywords/.
-->

miri64 commented:
Can we divide this up into two PRs please? One introducing the script (but not wired hot yet, just silently warning in the output) and one fixing the warnings?

Also, since this is doc, why isn't this included into `doccheck`?

aabadie commented:
> Can we divide this up into two PRs please?

Sure, that's what I wrote in the initial comment of this PR.

> since this is doc, why isn't this included into doccheck?

I thought it would be simpler to have it separate first. But I'm fine with merging the two once all grouping issues are fixed.

miri64 commented on doc.txt:
```diff
@@ -31,3 +31,13 @@
+
+/**
+ * @defgroup    tests Tests
+ * @brief       Tests applications
```
AFAIK those groups will be empty in the rendered doc. Do we really need them?

aabadie replied:
Yes, all examples are using `@ingroup examples` and tests `@ingroup tests`. I know this is not nice but at least it silents the check script.
There are 3 solutions:
1. Remove all `@ingroup tests` and `@ingroup examples`
2. Ignore undefined examples and tests groups in the check script
3. Define `tests` and `examples` groups and document them. If this solution is pushed a bit, we could have all examples and tests documented in doxygen (instead of READMEs).

I think `1.` should be avoided because it will touch too many files and maybe one day it can be useful to have those groups defined.
`2.` is the simplest.
`3.` is IMHO technically and from a centralized documentation the best solution. But requires more work.

miri64 replied:
I think 3 does make little sense. While I prefer API documentation and usage documentation to be centralized, the READMEs of applications should be kept *with the application* for two reasons:

1. One can easily open a README in command-line or their favorite editor/IDE while working with the example and does not need to open an extra browser window
2. The context might get lost if the README is moved somewhere else

As such, I prefer 2, but instead of ignoring the *groups* I rather would ignore the *directories* (as e.g. the vendor headers).

Furthermore, I don't see the point of being so strict with group names in .c-files, as they are not parsed by doxygen, so they can't break the rendered doc (which was the whole point of my initial issue), so it might also be sensible to ignore .c-files altogether. But maybe I'm in the minority with this opinion.

aabadie replied:
I dropped the defined examples and tests groups. Will update #9266 with solution 2.

miri64 commented on pkg/lwip/include/arch/cc.h:
```diff
@@ -8,7 +8,7 @@
 
 /**
  * @defgroup    lwip_arch_cc    Compiler and processor description
```
`pkg_lwip_arch_cc`?

aabadie replied:
Ok, but a bit out of the scope of this PR. If you insist, I can also change that.

miri64 replied:
It's less out of scope than the introduction of `auto_init` doc below since this PR is about group naming ;-).

aabadie replied:
Ok, you insist ;)

miri64 commented on pkg/lwip/include/arch/sys_arch.h:
```diff
@@ -8,7 +8,7 @@
 
 /**
  * @defgroup    lwip_arch_sys_arch Architecture depentent definitions
```
`pkg_lwip_arch_sys_arch`?

miri64 commented on pkg/lwip/include/lwipopts.h:
```diff
@@ -8,7 +8,7 @@
 
 /**
  * @defgroup    lwip_opts   lwIP options
```
`pkg_lwip_opts`?

aabadie replied:
all lwip pkg group are fixed

miri64 commented on sys/auto_init/auto_init.c:
```diff
@@ -7,7 +7,7 @@
  * directory for more details.
  *
- * @ingroup auto_init
+ * @ingroup sys_autoinit
```
Here and below: shouldn't it rather be `sys_auto_init`?

aabadie replied:
I kept the initially defined `sys_autoinit`, I'll use your suggestion.

aabadie replied:
name changed

miri64 commented on sys/include/auto_init.h:
```diff
@@ -8,22 +8,55 @@
+ * This module contains all netif drivers that support auto-initialization.
+ */
+
+/**
```
Unrelated change?

aabadie replied:
the changes related to auto_init was a bit problematic. I went for the improvement of the initial doc and add the missing groups there. Do you have a better suggestion ?

miri64 replied:
Ah sorry, forgot about this comment. Iff you insist to keep the doc improvement in this PR, please at least factor it out of https://github.com/RIOT-OS/RIOT/pull/9264/commits/6f91bf801bd3ede63fde923ea6fe4b027f4726cb into a separate commit.

aabadie replied:
> please at least factor it out of 6f91bf8 into a separate commit.

done

miri64 commented on sys/include/crypto/modes/ccm.h:
```diff
@@ -26,11 +27,16 @@
+#define CCM_ERR_INVALID_DATA_LENGTH         (-3)
+#define CCM_ERR_INVALID_LENGTH_ENCODING     (-4)
+#define CCM_ERR_INVALID_MAC_LENGTH          (-5)
+/** @} */
```
Unrelated

aabadie replied:
No, fixing the doxygen for this file is allowing the processing of doxygen in this file and, in the end, triggers a warning from doxygen because of undocumented defines. This change is fixing this. Do you want it in a separate commit ?

aabadie commented:
Rebased, now that #9266 is merged.

aabadie commented:
@miri64, are your concerns addressed here so I can squash ?

miri64 commented on sys/include/auto_init.h:
````diff
@@ -8,22 +8,55 @@
+ * This feature can be enabled in any application by adding the `auto_init`
+ * module to the application Makefile:
+ *
+ * ```mk
````
This syntax doesn't work in Doxygen. Try

```markdown
~~~~~~~~~~~~~~~~~~~~~~~ {.mk}
USEMODULE += auto_init
~~~~~~~~~~~~~~~~~~~~~~~
```

instead if you like to enable syntax highlighting, otherwise remove the `mk`.

aabadie replied:
removed the commit related to that change. Will open a dedicated PR for that later.

miri64 commented on sys/include/auto_init.h:
```diff
@@ -8,22 +8,55 @@
+ * `auto_init` initializes any included module that provides auto
+ * initialization capabilities.
+ *
+ * Drivers or cpu peripherals that provides a SAUL adaption and net interfaces
```
"network interfaces"

miri64 commented on sys/include/auto_init.h:
```diff
@@ -8,22 +8,55 @@
+ * initialization capabilities.
+ *
+ * Drivers or cpu peripherals that provides a SAUL adaption and net interfaces
+ * can be initialized automatically with auto_init.
```
`` around `auto_init`

miri64 commented on sys/include/auto_init.h:
```diff
@@ -8,22 +8,55 @@
+ * can be initialized automatically with auto_init.
+ *
+ * For high-level device drivers (@ref drivers), the default initialization
+ * parameters are taken from the `DRIVERS_params.h` files provided by the
```
`DRIVER`

miri64 commented on sys/include/auto_init.h:
```diff
@@ -8,22 +8,55 @@
+ * The modules will be initialized in the context of the main thread right
+ * before the main function gets called. Be aware that most modules expect to
+ * be initialized only once, so do not call a module's init function when using
+ * auto_init unless you know what you're doing.
```
```markdown
`auto_init`, unless you know what you're doing.
```

miri64 commented on sys/include/auto_init.h:
```diff
@@ -8,22 +8,55 @@
+ * @ingroup     sys_auto_init
+ * @brief       Provides auto-initialization of SAUL drivers.
+ *
+ * This module contains all high-level drivers (sensor or actuators) that
```
Please rephrase. This sentence makes little sense:

* What are high-level drivers? Do you mean SAUL as a high-level *interface* to the drivers?
* This module only allows for auto-initialization of those drivers using SAUL, and does not contain the drivers.

miri64 commented on sys/include/auto_init.h:
```diff
@@ -8,22 +8,55 @@
+ * @ingroup     sys_auto_init
+ * @brief       Provides auto-initialization of Netif drivers via GNRC.
+ *
+ * This module contains all netif drivers that support auto-initialization.
```
This module also doesn't contain any drivers, only enables their auto-initialization.

miri64 commented on sys/include/auto_init.h:
```diff
@@ -8,22 +8,55 @@
+/**
+ * @defgroup    sys_auto_init_gnrc_netif GNRC netif drivers auto-initialization
+ * @ingroup     sys_auto_init
+ * @brief       Provides auto-initialization of Netif drivers via GNRC.
```
`s/Netif/network device drivers/`. Ideally, link to `drivers_netdev`, e.g.

```markdown
Provides auto-initialization for [network device drivers](@ref drivers_netdev) via GNRC.
```

miri64 commented on sys/crypto/modes/cbc.c:
```diff
@@ -7,7 +7,7 @@
 
 /**
- * @ingroup     sys_crypto_modes
+ * @ingroup     sys_crypto_mode_cbc
```
If you just use `sys_crypto` here, the introduction of the new groups below would not be required.

aabadie replied:
You are right, fixed now.

miri64 commented on sys/include/crypto/modes/cbc.h:
```diff
@@ -7,11 +7,12 @@
  */
 
 /**
+ * @defgroup    sys_crypto_mode_cbc Cipher block chaining mode
```
This is really completely unrelated. These files are perfectly fine to all be grouped in `sys_crypto`. If you disagree, please move this group introductions to a separate PR.

aabadie commented:
@miri64, before this PR becomes unmaintainable, do you have other concerns ? I think I addressed all your comments and moved out the unrelated things.

aabadie commented:
squashed, let's wait for the CI.

miri64 commented:
Travis fail is unrelated (and due to an old cppcheck version, I believe) => go
